// @trpc/server@11.1.5-canary.1/dist/unstable-core-do-not-import.js
'use strict';

var createProxy = require('./unstable-core-do-not-import/createProxy.js');
var formatter = require('./unstable-core-do-not-import/error/formatter.js');
var getErrorShape = require('./unstable-core-do-not-import/error/getErrorShape.js');
var TRPCError = require('./unstable-core-do-not-import/error/TRPCError.js');
var contentType = require('./unstable-core-do-not-import/http/contentType.js');
var contentTypeParsers = require('./unstable-core-do-not-import/http/contentTypeParsers.js');
var formDataToObject = require('./unstable-core-do-not-import/http/formDataToObject.js');
var getHTTPStatusCode = require('./unstable-core-do-not-import/http/getHTTPStatusCode.js');
var abortError = require('./unstable-core-do-not-import/http/abortError.js');
var parseConnectionParams = require('./unstable-core-do-not-import/http/parseConnectionParams.js');
var resolveResponse = require('./unstable-core-do-not-import/http/resolveResponse.js');
var initTRPC = require('./unstable-core-do-not-import/initTRPC.js');
var middleware = require('./unstable-core-do-not-import/middleware.js');
var parser = require('./unstable-core-do-not-import/parser.js');
var procedure = require('./unstable-core-do-not-import/procedure.js');
var procedureBuilder = require('./unstable-core-do-not-import/procedureBuilder.js');
var rootConfig = require('./unstable-core-do-not-import/rootConfig.js');
var router = require('./unstable-core-do-not-import/router.js');
var codes = require('./unstable-core-do-not-import/rpc/codes.js');
var parseTRPCMessage = require('./unstable-core-do-not-import/rpc/parseTRPCMessage.js');
var jsonl = require('./unstable-core-do-not-import/stream/jsonl.js');
var sse = require('./unstable-core-do-not-import/stream/sse.js');
var tracked = require('./unstable-core-do-not-import/stream/tracked.js');
var createDeferred = require('./unstable-core-do-not-import/stream/utils/createDeferred.js');
var disposable = require('./unstable-core-do-not-import/stream/utils/disposable.js');
var asyncIterable = require('./unstable-core-do-not-import/stream/utils/asyncIterable.js');
var transformer = require('./unstable-core-do-not-import/transformer.js');
var utils = require('./unstable-core-do-not-import/utils.js');
var error = require('./vendor/standard-schema-v1/error.js');
var unpromise = require('./vendor/unpromise/unpromise.js');



exports.createFlatProxy = createProxy.createFlatProxy;
exports.createRecursiveProxy = createProxy.createRecursiveProxy;
exports.defaultFormatter = formatter.defaultFormatter;
exports.getErrorShape = getErrorShape.getErrorShape;
exports.TRPCError = TRPCError.TRPCError;
exports.getCauseFromUnknown = TRPCError.getCauseFromUnknown;
exports.getTRPCErrorFromUnknown = TRPCError.getTRPCErrorFromUnknown;
exports.getRequestInfo = contentType.getRequestInfo;
exports.octetInputParser = contentTypeParsers.octetInputParser;
exports.formDataToObject = formDataToObject.formDataToObject;
exports.HTTP_CODE_TO_JSONRPC2 = getHTTPStatusCode.HTTP_CODE_TO_JSONRPC2;
exports.JSONRPC2_TO_HTTP_CODE = getHTTPStatusCode.JSONRPC2_TO_HTTP_CODE;
exports.getHTTPStatusCode = getHTTPStatusCode.getHTTPStatusCode;
exports.getHTTPStatusCodeFromError = getHTTPStatusCode.getHTTPStatusCodeFromError;
exports.getStatusCodeFromKey = getHTTPStatusCode.getStatusCodeFromKey;
exports.getStatusKeyFromCode = getHTTPStatusCode.getStatusKeyFromCode;
exports.isAbortError = abortError.isAbortError;
exports.throwAbortError = abortError.throwAbortError;
exports.parseConnectionParamsFromString = parseConnectionParams.parseConnectionParamsFromString;
exports.parseConnectionParamsFromUnknown = parseConnectionParams.parseConnectionParamsFromUnknown;
exports.resolveResponse = resolveResponse.resolveResponse;
exports.initTRPC = initTRPC.initTRPC;
exports.createInputMiddleware = middleware.createInputMiddleware;
exports.createMiddlewareFactory = middleware.createMiddlewareFactory;
exports.createOutputMiddleware = middleware.createOutputMiddleware;
exports.experimental_standaloneMiddleware = middleware.experimental_standaloneMiddleware;
exports.middlewareMarker = middleware.middlewareMarker;
exports.getParseFn = parser.getParseFn;
exports.procedureTypes = procedure.procedureTypes;
exports.createBuilder = procedureBuilder.createBuilder;
exports.isServerDefault = rootConfig.isServerDefault;
exports.callProcedure = router.callProcedure;
exports.createCallerFactory = router.createCallerFactory;
exports.createRouterFactory = router.createRouterFactory;
exports.getProcedureAtPath = router.getProcedureAtPath;
exports.lazy = router.lazy;
exports.mergeRouters = router.mergeRouters;
exports.TRPC_ERROR_CODES_BY_KEY = codes.TRPC_ERROR_CODES_BY_KEY;
exports.TRPC_ERROR_CODES_BY_NUMBER = codes.TRPC_ERROR_CODES_BY_NUMBER;
exports.parseTRPCMessage = parseTRPCMessage.parseTRPCMessage;
exports.isPromise = jsonl.isPromise;
exports.jsonlStreamConsumer = jsonl.jsonlStreamConsumer;
exports.jsonlStreamProducer = jsonl.jsonlStreamProducer;
exports.sseHeaders = sse.sseHeaders;
exports.sseStreamConsumer = sse.sseStreamConsumer;
exports.sseStreamProducer = sse.sseStreamProducer;
exports.isTrackedEnvelope = tracked.isTrackedEnvelope;
exports.sse = tracked.sse;
exports.tracked = tracked.tracked;
exports.createDeferred = createDeferred.createDeferred;
exports.makeAsyncResource = disposable.makeAsyncResource;
exports.makeResource = disposable.makeResource;
exports.iteratorResource = asyncIterable.iteratorResource;
exports.takeWithGrace = asyncIterable.takeWithGrace;
exports.withMaxDuration = asyncIterable.withMaxDuration;
exports.defaultTransformer = transformer.defaultTransformer;
exports.getDataTransformer = transformer.getDataTransformer;
exports.transformResult = transformer.transformResult;
exports.transformTRPCResponse = transformer.transformTRPCResponse;
exports.abortSignalsAnyPonyfill = utils.abortSignalsAnyPonyfill;
exports.assert = utils.assert;
exports.identity = utils.identity;
exports.isAsyncIterable = utils.isAsyncIterable;
exports.isFunction = utils.isFunction;
exports.isObject = utils.isObject;
exports.mergeWithoutOverrides = utils.mergeWithoutOverrides;
exports.noop = utils.noop;
exports.omitPrototype = utils.omitPrototype;
exports.run = utils.run;
exports.sleep = utils.sleep;
exports.unsetMarker = utils.unsetMarker;
exports.StandardSchemaV1Error = error.StandardSchemaV1Error;
exports.Unpromise = unpromise.Unpromise;
